homeassistant/components/hassio/auth.py

```diff
@@ -0,0 +1,67 @@
+"""Implement the auth feature from Hass.io for Add-ons."""
+import logging
+from ipaddress import ip_address
+import os
+
+from aiohttp import web
+from aiohttp.web_exceptions import (
+    HTTPForbidden, HTTPNotFound, HTTPUnauthorized)
+
+from homeassistant.core import callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.components.http import HomeAssistantView
+from homeassistant.components.http.const import KEY_REAL_IP
+
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_USERNAME = 'username'
+ATTR_PASSWORD = 'password'
+
+
+@callback
+def async_setup_auth(hass):
+    """Auth setup."""
+    hassio_auth = HassIOAuth(hass)
+    hass.http.register_view(hassio_auth)
+
+
+class HassIOAuth(HomeAssistantView):
+    """Hass.io view to handle base part."""
+
+    name = "api:hassio_auth"
+    url = "/api/hassio_auth"
+
+    def __init__(self, hass):
+        """Initialize WebView."""
+        self.hass = hass
+
+    async def post(self, request):
```

Comment:
Wrap this in `@RequestDataValidator(vol.Schema({ … })))`

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/http/data_validator.py#L11